### PR TITLE
WIP: Use tasks for ConstraintBuilder multithreading

### PR DESCRIPTION
**Pull request by gaschler**
This is work in progress not ready to commit.
ConstraintBuilder3D is not yet integrated but should be analogous.

Please only give high-level comments. After that, I can provide smaller PRs.

### CODE REVIEW

File: cartographer/common/thread_pool_test.cc [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+  ThreadPool pool(2);
+  Receiver receiver;
+  auto task_2 = common::make_unique<Task>();
+  task_2->SetWorkItem([&receiver]() { receiver.Receive(2); });
+  {
+    auto task_1 = common::make_unique<Task>();
+    task_1->SetWorkItem([&receiver]() { receiver.Receive(1); });
+    auto weak_task_1 = pool.ScheduleWhenReady(std::move(task_1));
+    task_2->AddDependency(weak_task_1);
+    // task_1 goes out of scope.
```
Comment (pifon2a) [inline cartographer/common/thread_pool_test.cc]: I do not think that this comment is needed.

File: cartographer/mapping/internal/constraints/constraint_builder_2d.cc [hunk 2]
```diff
@@ -258,50 +266,41 @@ void ConstraintBuilder2D::ComputeConstraint(
...
-          LOG(INFO) << constraints_.size() << " computations resulted in "
-                    << result.size() << " additional constraints.";
-          LOG(INFO) << "Score histogram:\n" << score_histogram_.ToString(10);
-        }
-        constraints_.clear();
-        callback = std::move(when_done_);
-        when_done_.reset();
+    CHECK(when_done_ != nullptr);
+    for (const std::unique_ptr<Constraint>& constraint : constraints_) {
+      if (constraint != nullptr) {
```
Comment (pifon2a) [inline cartographer/mapping/internal/constraints/constraint_builder_2d.cc]: nit:
```
if (constraint == nullptr) continue;
result.push_back(*constraint);
```

File: cartographer/mapping/internal/constraints/constraint_builder_2d.cc [hunk 3]
```diff
@@ -77,44 +79,68 @@ void ConstraintBuilder2D::MaybeAddConstraint(
...
-    ++pending_computations_[current_computation_];
-    const int current_computation = current_computation_;
-    ScheduleSubmapScanMatcherConstructionAndQueueWorkItem(
-        submap_id, submap->grid(), [=]() EXCLUDES(mutex_) {
-          ComputeConstraint(submap_id, submap, node_id,
-                            false, /* match_full_submap */
-                            constant_data, initial_relative_pose, constraint);
-          FinishComputation(current_computation);
-        });
+  if (!sampler_.Pulse()) return;
```
Comment (pifon2a) [inline cartographer/mapping/internal/constraints/constraint_builder_2d.cc]: good!

File: cartographer/common/thread_pool_test.cc [hunk 4]
```diff
@@ -0,0 +1,130 @@
...
+      have_enough_numbers = locker.AwaitWithTimeout(
+          [this, &expected_numbers]() REQUIRES(mutex_) {
+            return (received_numbers_.size() >= expected_numbers.size());
+          },
+          common::FromSeconds(0.1));
+    }
+    EXPECT_EQ(expected_numbers, received_numbers_);
+  }
+
+  std::list<int> received_numbers_;
```
Comment (pifon2a) [inline cartographer/common/thread_pool_test.cc]: std::vector or std::deque?

File: cartographer/common/internal/testing/thread_pool_for_testing.h [hunk 5]
```diff
@@ -43,7 +46,8 @@ class ThreadPoolForTesting : public ThreadPoolInterface {
   std::thread thread_ GUARDED_BY(mutex_);
   bool running_ GUARDED_BY(mutex_) = true;
   bool idle_ GUARDED_BY(mutex_) = true;
-  std::deque<std::function<void()>> work_queue_ GUARDED_BY(mutex_);
+  std::deque<std::shared_ptr<Task>> task_queue_ GUARDED_BY(mutex_);
```
Comment (pifon2a) [inline cartographer/common/internal/testing/thread_pool_for_testing.h]: queue?

File: cartographer/common/thread_pool.h [hunk 6]
```diff
@@ -45,15 +57,24 @@ class ThreadPool : public ThreadPoolInterface {
...
+
  private:
   void DoWork();
 
   Mutex mutex_;
   bool running_ GUARDED_BY(mutex_) = true;
   std::vector<std::thread> pool_ GUARDED_BY(mutex_);
-  std::deque<std::function<void()>> work_queue_ GUARDED_BY(mutex_);
+  std::deque<std::shared_ptr<Task>> task_queue_ GUARDED_BY(mutex_);
+  std::map<Task*, std::shared_ptr<Task>> tasks_not_ready_ GUARDED_BY(mutex_);
```
Comment (pifon2a) [inline cartographer/common/thread_pool.h]: unordered_map? You do not iterate over it anywhere I think.

File: cartographer/common/task.h [hunk 7]
```diff
@@ -0,0 +1,73 @@
...
+namespace cartographer {
+namespace common {
+
+class ThreadPoolInterface;
+
+class Task {
+ public:
+  using WorkItem = std::function<void()>;
+  enum State { NEW, DISPATCHED, READY, RUNNING, COMPLETED };
+
```
Comment (pifon2a) [inline cartographer/common/task.h]: Task() = default?

File: cartographer/mapping/internal/constraints/constraint_builder_2d.cc [hunk 8]
```diff
@@ -258,50 +266,41 @@ void ConstraintBuilder2D::ComputeConstraint(
...
+  (*callback)(result);
 }
 
 int ConstraintBuilder2D::GetNumFinishedNodes() {
   common::MutexLocker locker(&mutex_);
-  if (pending_computations_.empty()) {
-    return current_computation_;
-  }
-  return pending_computations_.begin()->first;
+  return num_finished_nodes_;
```
Comment (cschuet) [inline cartographer/mapping/internal/constraints/constraint_builder_2d.cc]: It's a shame to take a lock for something so trivial like this counter. Let's turn this guy into a std::atomic

File: cartographer/mapping/internal/constraints/constraint_builder_2d.cc [hunk 9]
```diff
@@ -77,44 +79,68 @@ void ConstraintBuilder2D::MaybeAddConstraint(
...
+  finish_node_task_->AddDependency(weak_task);
 }
 
 void ConstraintBuilder2D::NotifyEndOfNode() {
   common::MutexLocker locker(&mutex_);
-  ++current_computation_;
+  CHECK(finish_node_task_ != nullptr);
+  finish_node_task_->SetWorkItem([this] {
+    common::MutexLocker locker(&mutex_);
+    ++num_finished_nodes_;
```
Comment (cschuet) [inline cartographer/mapping/internal/constraints/constraint_builder_2d.cc]: Make atomic and don't take lock.

File: cartographer/mapping/internal/constraints/constraint_builder_2d.h [hunk 10]
```diff
@@ -146,28 +139,25 @@ class ConstraintBuilder2D {
   std::unique_ptr<std::function<void(const Result&)>> when_done_
       GUARDED_BY(mutex_);
 
-  // Index of the node in reaction to which computations are currently
-  // added. This is always the highest node index seen so far, even when older
+  // Number of the node in reaction to which computations are currently
+  // added. This is always the number of nodes seen so far, even when older
   // nodes are matched against a new submap.
-  int current_computation_ GUARDED_BY(mutex_) = 0;
+  int num_started_nodes_ GUARDED_BY(mutex_) = 0;
```
Comment (cschuet) [inline cartographer/mapping/internal/constraints/constraint_builder_2d.h]: make atomic and don't take lock

File: cartographer/mapping/internal/constraints/constraint_builder_2d.h [hunk 11]
```diff
@@ -146,28 +139,25 @@ class ConstraintBuilder2D {
...
 
-  // For each added node, maps to the number of pending computations that were
-  // added for it.
-  std::map<int, int> pending_computations_ GUARDED_BY(mutex_);
+  int num_finished_nodes_ GUARDED_BY(mutex_) = 0;
```
Comment (cschuet) [inline cartographer/mapping/internal/constraints/constraint_builder_2d.h]: make atomic and don't take lock

File: cartographer/mapping/internal/constraints/constraint_builder_2d.cc [hunk 12]
```diff
@@ -77,44 +79,68 @@ void ConstraintBuilder2D::MaybeAddConstraint(
...
+  });
+  CHECK(when_done_task_ != nullptr);
+  auto weak_finish_node_task =
+      thread_pool_->ScheduleWhenReady(std::move(finish_node_task_));
+  finish_node_task_ = common::make_unique<common::Task>();
+  when_done_task_->AddDependency(weak_finish_node_task);
+  ++num_started_nodes_;
```
Comment (cschuet) [inline cartographer/mapping/internal/constraints/constraint_builder_2d.cc]: Who is even reading this guy? I only see a write.

File: cartographer/mapping/internal/constraints/constraint_builder_2d.cc [hunk 13]
```diff
@@ -123,60 +149,42 @@ void ConstraintBuilder2D::WhenDone(
...
                                                      const Grid2D* const grid) {
-  auto submap_scan_matcher =
-      common::make_unique<scan_matching::FastCorrelativeScanMatcher2D>(
-          *grid, options_.fast_correlative_scan_matcher_options());
-  common::MutexLocker locker(&mutex_);
-  submap_scan_matchers_[submap_id] = {grid, std::move(submap_scan_matcher)};
-  for (const std::function<void()>& work_item :
-       submap_queued_work_items_[submap_id]) {
-    thread_pool_->Schedule(work_item);
+  if (submap_scan_matchers_.count(submap_id) == 0) {
```
Comment (cschuet) [inline cartographer/mapping/internal/constraints/constraint_builder_2d.cc]: Hmm maybe
```
if (submap_scan_matchers_.count(submap_id) != 0) {
  return &submap_scan_matchers_.at(submap_id);
}
...
```

**Comment by gaschler** [comment]
Here are the results of the performance comparison:

Elapsed wall time [s] over 3 trials | master@3dd37d | tasks-2@352d07
-- | -- | --
b0-2014-07-11-10-58-16.bag | 28.84 ±0.4 | 29.35 ±0.7
b2-2015-04-28-13-01-40.bag | 117.98 ±0.5 | 123.36 ±4


(Run without rviz, i.e. `catkin_make_isolated --use-ninja && source devel_isolated/setup.bash && roslaunch cartographer_ros offline_backpack_2d.launch no_rviz:=true bag_filenames:=${HOME}/Downloads/b0-2014-07-11-10-58-16.bag`)

File: cartographer/mapping/internal/constraints/constraint_builder_2d.cc (hunk 13 above)
Comment (gaschler) [reply cartographer/mapping/internal/constraints/constraint_builder_2d.cc]: Done

File: cartographer/mapping/internal/constraints/constraint_builder_2d.h (hunk 11 above)
Comment (gaschler) [reply cartographer/mapping/internal/constraints/constraint_builder_2d.h]: Added a todo.

File: cartographer/mapping/internal/constraints/constraint_builder_2d.h (hunk 10 above)
Comment (gaschler) [reply cartographer/mapping/internal/constraints/constraint_builder_2d.h]: Added a todo.

File: cartographer/mapping/internal/constraints/constraint_builder_2d.cc (hunk 9 above)
Comment (gaschler) [reply cartographer/mapping/internal/constraints/constraint_builder_2d.cc]: Added a todo.

File: cartographer/mapping/internal/constraints/constraint_builder_2d.cc (hunk 8 above)
Comment (gaschler) [reply cartographer/mapping/internal/constraints/constraint_builder_2d.cc]: Added a todo.

File: cartographer/mapping/internal/constraints/constraint_builder_2d.cc (hunk 2 above)
[the remark this answers is not in the document]
Comment (gaschler) [reply cartographer/mapping/internal/constraints/constraint_builder_2d.cc]: Done, added a comment to where constraints_ is declared. (nullptr signals that the score was below threshold).

File: cartographer/mapping/internal/constraints/constraint_builder_2d.cc [hunk 14]
```diff
@@ -258,50 +266,41 @@ void ConstraintBuilder2D::ComputeConstraint(
   }
 }
 
-void ConstraintBuilder2D::FinishComputation(const int computation_index) {
+void ConstraintBuilder2D::RunCallback() {
```
[the remark this answers is not in the document]
Comment (gaschler) [reply cartographer/mapping/internal/constraints/constraint_builder_2d.cc]: Done

File: cartographer/mapping/internal/constraints/constraint_builder_2d.cc [hunk 15]
```diff
@@ -77,44 +79,68 @@ void ConstraintBuilder2D::MaybeAddConstraint(
...
 void ConstraintBuilder2D::NotifyEndOfNode() {
   common::MutexLocker locker(&mutex_);
-  ++current_computation_;
+  CHECK(finish_node_task_ != nullptr);
+  finish_node_task_->SetWorkItem([this] {
+    common::MutexLocker locker(&mutex_);
+    ++num_finished_nodes_;
+  });
+  CHECK(when_done_task_ != nullptr);
+  auto weak_finish_node_task =
```
[the remark this answers is not in the document]
Comment (gaschler) [reply cartographer/mapping/internal/constraints/constraint_builder_2d.cc]: Done

File: cartographer/mapping/internal/constraints/constraint_builder_2d.cc [hunk 16]
```diff
@@ -77,44 +79,68 @@ void ConstraintBuilder2D::MaybeAddConstraint(
...
+  task->AddDependency(scan_matcher->scan_matcher_factory_task);
+  auto weak_task = thread_pool_->ScheduleWhenReady(std::move(task));
+  CHECK(finish_node_task_ != nullptr);
+  finish_node_task_->AddDependency(weak_task);
 }
 
 void ConstraintBuilder2D::NotifyEndOfNode() {
   common::MutexLocker locker(&mutex_);
-  ++current_computation_;
+  CHECK(finish_node_task_ != nullptr);
```
[the remark this answers is not in the document]
Comment (gaschler) [reply cartographer/mapping/internal/constraints/constraint_builder_2d.cc]: Done

File: cartographer/mapping/internal/constraints/constraint_builder_2d.cc [hunk 17]
```diff
@@ -77,44 +79,68 @@ void ConstraintBuilder2D::MaybeAddConstraint(
...
+  const auto* scan_matcher =
+      DispatchScanMatcherConstruction(submap_id, submap->grid());
+  auto task = common::make_unique<common::Task>();
+  task->SetWorkItem([=]() EXCLUDES(mutex_) {
+    ComputeConstraint(submap_id, submap, node_id, true, /* match_full_submap */
+                      constant_data, transform::Rigid2d::Identity(),
+                      *scan_matcher, constraint);
+  });
+  task->AddDependency(scan_matcher->scan_matcher_factory_task);
+  auto weak_task = thread_pool_->ScheduleWhenReady(std::move(task));
```
[the remark this answers is not in the document]
Comment (gaschler) [reply cartographer/mapping/internal/constraints/constraint_builder_2d.cc]: Done

File: cartographer/mapping/internal/constraints/constraint_builder_2d.cc [hunk 18]
```diff
@@ -77,44 +79,68 @@ void ConstraintBuilder2D::MaybeAddConstraint(
...
-  ScheduleSubmapScanMatcherConstructionAndQueueWorkItem(
-      submap_id, submap->grid(), [=]() EXCLUDES(mutex_) {
-        ComputeConstraint(
-            submap_id, submap, node_id, true, /* match_full_submap */
-            constant_data, transform::Rigid2d::Identity(), constraint);
-        FinishComputation(current_computation);
-      });
+  const auto* scan_matcher =
+      DispatchScanMatcherConstruction(submap_id, submap->grid());
+  auto task = common::make_unique<common::Task>();
```
[the remark this answers is not in the document]
Comment (gaschler) [reply cartographer/mapping/internal/constraints/constraint_builder_2d.cc]: Done

File: cartographer/mapping/internal/constraints/constraint_builder_2d.cc [hunk 19]
```diff
@@ -123,60 +149,42 @@ void ConstraintBuilder2D::WhenDone(
   CHECK(when_done_ == nullptr);
   when_done_ =
       common::make_unique<std::function<void(const Result&)>>(callback);
```
[the remark this answers is not in the document]
Comment (gaschler) [reply cartographer/mapping/internal/constraints/constraint_builder_2d.cc]: Added a TODO

File: cartographer/mapping/internal/constraints/constraint_builder_2d.cc [hunk 20]
```diff
@@ -77,44 +79,68 @@ void ConstraintBuilder2D::MaybeAddConstraint(
...
+      DispatchScanMatcherConstruction(submap_id, submap->grid());
+  auto task = common::make_unique<common::Task>();
+  task->SetWorkItem([=]() EXCLUDES(mutex_) {
+    ComputeConstraint(submap_id, submap, node_id, false, /* match_full_submap */
+                      constant_data, initial_relative_pose, *scan_matcher,
+                      constraint);
+  });
+  task->AddDependency(scan_matcher->scan_matcher_factory_task);
+  auto weak_task = thread_pool_->ScheduleWhenReady(std::move(task));
+  CHECK(finish_node_task_ != nullptr);
```
[the remark this answers is not in the document]
Comment (gaschler) [reply cartographer/mapping/internal/constraints/constraint_builder_2d.cc]: Done

File: cartographer/mapping/internal/constraints/constraint_builder_2d.cc [hunk 21]
```diff
@@ -77,44 +79,68 @@ void ConstraintBuilder2D::MaybeAddConstraint(
...
+  if (when_done_) {
+    LOG(WARNING)
+        << "MaybeAddConstraint was called while WhenDone was scheduled.";
   }
+  constraints_.emplace_back();
+  kQueueLengthMetric->Set(constraints_.size());
+  auto* const constraint = &constraints_.back();
+  const auto* scan_matcher =
+      DispatchScanMatcherConstruction(submap_id, submap->grid());
+  auto task = common::make_unique<common::Task>();
```
[the remark this answers is not in the document]
Comment (gaschler) [reply cartographer/mapping/internal/constraints/constraint_builder_2d.cc]: Done

File: cartographer/mapping/internal/constraints/constraint_builder_2d.cc [hunk 22]
```diff
@@ -77,44 +79,68 @@ void ConstraintBuilder2D::MaybeAddConstraint(
...
+  task->SetWorkItem([=]() EXCLUDES(mutex_) {
+    ComputeConstraint(submap_id, submap, node_id, false, /* match_full_submap */
+                      constant_data, initial_relative_pose, *scan_matcher,
+                      constraint);
+  });
+  task->AddDependency(scan_matcher->scan_matcher_factory_task);
+  auto weak_task = thread_pool_->ScheduleWhenReady(std::move(task));
```
[the remark this answers is not in the document]
Comment (gaschler) [reply cartographer/mapping/internal/constraints/constraint_builder_2d.cc]: Done

File: cartographer/common/task.h (hunk 7 above)
Comment (gaschler) [reply cartographer/common/task.h]: Done

File: cartographer/common/thread_pool.h (hunk 6 above)
Comment (gaschler) [reply cartographer/common/thread_pool.h]: Done

File: cartographer/common/thread_pool_test.cc (hunk 4 above)
Comment (gaschler) [reply cartographer/common/thread_pool_test.cc]: Done, std::vector.

File: cartographer/mapping/internal/constraints/constraint_builder_2d.cc (hunk 2 above)
Comment (gaschler) [reply cartographer/mapping/internal/constraints/constraint_builder_2d.cc]: Done

File: cartographer/common/thread_pool_test.cc (hunk 1 above)
Comment (gaschler) [reply cartographer/common/thread_pool_test.cc]: Done

**Comment by gaschler** [comment]
Thanks for your helpful reviews.
The actual change will be made in #1066 #1118 #1119 (taking your comments into account), so I'm closing this one.

**Pull request closed without merging**